streaming search that aggregates and ranks for FlushWallTime

keegancsmith opened the pull request:
We introduce FlushWallTime which will adjust streaming searches behaviour to collect results for that duration of time before sending. This gives a chance for zoekt to do ranking, while still maintaining low latency for time to first result.

Additionally we also make it so that streaming search respects MaxDocDisplayCount.

Test Plan: go test

stefanhengl commented:
I wonder whether this logic should sit at the network layer? Streaming is the only stakeholder of `Progress`. At the same time we wouldn't have to introduce the complexity in shards.go

keegancsmith commented:
> I wonder whether this logic should sit at the network layer? Streaming is the only stakeholder of `Progress`. At the same time we wouldn't have to introduce the complexity in shards.go

We wanna do ranking and truncation before calling copyFiles, which is why it ends up here. I tried to isolate it into something understable and decoupled (the stuff in aggregate.go), but I'm happy to improve design/code so it feels understandable / well tested.

keegancsmith commented:
Not the best place to keep this info, but here is the raw notes from the call were we discussed this

#### weird stuff we noticed

Progress results are not actually streaming out. This is because at the network layer we suppress events with no filematches. This is a happy accident, because our very first event has no progress (just the stat wait time). We should improve stuff so we always set Progress. We should probably move the decision logic for periodic progress events into the network layer.

`resultCountFactor` was a heuristic based on sourcegraph from many years ago that was never tweaked. We had a small corpus and only one zoekt. We should experiment with it / throw it away.

Our `userProbablyWantsToWaitLonger` logic is broken. If you are going via streaming, it will always trigger to true due to using the wrong default limit. We shouldn't even decide this this way, the layers above this know if the user tweaked the count. Worse than this, it completely blows up the k value we calculate in `resultCountFactor`. This blows up the amount of work zoekt is doing.

#### actually about ranking

need to ensure Score takes into account Repo rank

this is flagged by the size of `count:`, because our memory use will be affected by the number of results we need to send (ie the size of our priority queue). Additionally when someone asks for large counts, they shouldn't care about ranking they care about stats or exhaustiveness.

-   flush means rank what you have (and hopefully truncate) and send it.
-   Each zoekt will flush after 300ms or sooner
-   Frontend will flush after 500ms or sooner
-   what do we do when one zoekt overwhelms the queue?

If count is small enough, we should set MaxDocDisplayCount to count. Then no zoekt will send more than count.

Should the flushing be periodic? Otherwise the user won't see progress for exhaustive/slow searches.

Should we be thinking about determinism for larger result sets?

frontend sort and truncate = mem is bounded by 49 \* count heap = mem is bounded by count

the way our clients work is that we will have unmarshalled all of the results anyways, so heap in practice won't actually save memory. So the simple impl of sort and slice is fine.

If we want to use this algorithm with large count, we likely need another parameter around how many results zoekt will buffer before flushing. This is like MaxDocDisplayCount, except I guess you call it MaxDocFlushCount. Or after the first flush we revert back to streaming and rely on frontend to do ranking.

keegancsmith committed:
shards: introduce SearchOptions.FlushWallTime

This option makes it so that streaming search will collect results for
FlushWallTime before sending. This allows a client to provide
best-effort ranking based on time. We hope this will be a sweet spot
between the benefits of streaming latency and relevance of results.

Test Plan: go test. TODO add unit tests exercising this specific logic.

keegancsmith commented:
@stefanhengl i've updated this PR so that it is rebased and cleaned up. i still want to add some tests before landing, but this feels like something good to use for ranking across repos in frontend. We may wanna export `newFlushCollectSender` since I imagine it could also be used directly by frontend.

stefanhengl committed:
port changes from frontend, update searchOptions

stefanhengl committed:
use stable sort

stefanhengl committed:
better indentation

stefanhengl committed:
remove blank

stefanhengl committed:
wrapping, wrapping

stefanhengl committed:
remove maxSizeBytes

keegancsmith commented on api.go:
```diff
@@ -777,6 +777,15 @@ type SearchOptions struct {
+
+	// MaxSizesBytes if non-zero limits the number of bytes Zoekt is allowed to hold
+	// in memory for file matches before flushing.
+	MaxSizeBytes int
```
need to remove, will follow up in another PR

keegancsmith commented on api.go:
```diff
@@ -86,7 +86,7 @@ type FileMatch struct {
 }
 
-func (m *FileMatch) sizeBytes() (sz uint64) {
+func (m *FileMatch) SizeBytes() (sz uint64) {
```
We don't need to export this anymore.

stefanhengl committed:
unexport SizeBytes
